### PR TITLE
Align KTO with DPO: Support VLM

### PR DESCRIPTION
**Pull request by albertvillanova**
Align KTO with DPO: Support VLM.

Part of:
- #4786

This PR introduces support for vision-language models (VLMs) and multimodal datasets to the KTOTrainer. The main changes include a new data collator for unpaired vision preference data, dynamic detection of vision datasets, and updates throughout the trainer to handle multimodal inputs, on-the-fly processing, and proper model argument handling.

### Solution

New class `DataCollatorForVisionUnpairedPreference` mirrors `DataCollatorForVisionPreference` from DPO but adapted for KTO's unpaired structure (one completion per example instead of a paired chosen/rejected).

Key design decision: the KL divergence term (which in the text-only path is pre-computed as a shifted `KL_completion_ids` column during dataset preparation) is handled **inside the collator** for VLMs. This keeps the vision dataset fully raw (no pre-processing columns), consistent with how DPO's vision collator works. The collator cycles completions by +1 within each batch (`completions[-1:] + completions[:-1]`), the same mismatching strategy as `_get_kl_completion_ids`, and builds the `KL_completion_*` tensors on the fly alongside the main completion tensors. Processor calls for `token_type_ids` (Gemma) and `mm_token_type_ids` (Qwen2.5-VL) are handled separately for the main and KL sequences since they can have different lengths.

### Changes

**Vision-Language Model and Multimodal Dataset Support**

* Added `DataCollatorForVisionUnpairedPreference` to handle on-the-fly tokenization and image processing for vision datasets, including KL sequence generation and multimodal message preparation.
* Updated the trainer to automatically detect vision datasets and select the appropriate collator, raising errors for incompatible model/dataset combinations or unsupported settings. 

**On-the-fly Data Processing and Collation**

* Skipped dataset pre-processing for vision datasets, moving all tokenization and image handling into the collator for VLMs.
* Set expected batch signature columns for vision datasets to include image and multimodal fields.

**Model Forward Pass and KL Handling**

* Updated model input construction for both main and KL batches to support vision-language inputs (e.g., `pixel_values`, `image_grid_thw`, `token_type_ids`, `mm_token_type_ids`).

**Utility and Import Changes**

* Added imports for new multimodal utilities and the `flush_left` function to support input formatting and collation. 

These changes collectively enable seamless training of vision-language models on preference datasets containing both text and images.

---

> [!NOTE]
> **Medium Risk**
> Touches core KTO training, KL pairing, and model forward paths for VLMs; mis-wired multimodal kwargs or backbone choice could break VLM training while text paths largely stay behind dataset detection.
> 
> **Overview**
> Adds **vision-language model (VLM) support** to `KTOTrainer`, aligned with DPO’s multimodal path.
> 
> Introduces **`DataCollatorForVisionUnpairedPreference`**, which tokenizes and processes images at batch time from raw `prompt` / `completion` / `image(s)` / `label` rows. For KTO’s KL term it **cycles completions within the batch** (same idea as text `_get_kl_completion_ids`) and emits `KL_completion_*` tensors plus Gemma/Qwen **`token_type_ids` / `mm_token_type_ids`** handling and **`flush_left`** padding.
> 
> **Trainer wiring:** detects vision datasets and `ProcessorMixin` (`_is_vlm`), picks the vision collator, **skips `_prepare_dataset`** for VLMs, and updates **signature columns** for raw multimodal fields. **Guards** block vision + non-VLM, `precompute_ref_log_probs`, and paired `chosen`/`rejected` formats.
> 
> **Forward passes** build `model_kwargs` from the full collator batch (e.g. `pixel_values`) instead of only `input_ids`/`attention_mask`; KL uses renamed KL tensors and KL-specific token-type fields. **Liger path** uses VLM **`base_model` / `.model`** (not `get_decoder()`) so vision tokens are injected correctly on older transformers.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 261dad4e2a79b17e812d78b472f7616971e56964. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/kto/kto_trainer.py [hunk 1]
```diff
@@ -930,6 +1167,15 @@ def _compute_kl_logps(self, model, batch):
...
             }
+            # VLM keys: pixel_values and image_grid_thw are shared with the main completion (same prompt+image).
+            # token_type_ids and mm_token_type_ids are KL-specific because the sequence length differs.
+            for key in ["pixel_values", "image_grid_thw", "image_position_ids"]:
+                if key in batch:
+                    KL_model_kwargs[key] = batch[key]
+            if "KL_completion_token_type_ids" in batch:
+                KL_model_kwargs["token_type_ids"] = batch["KL_completion_token_type_ids"]
+            if "KL_completion_mm_token_type_ids" in batch:
+                KL_model_kwargs["mm_token_type_ids"] = batch["KL_completion_mm_token_type_ids"]
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py:1153]: DPO's Liger path has the exact same issue: I'm addressing that in a separate PR.

[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py:1153]: See:
- https://github.com/huggingface/trl/pull/5943

File: trl/experimental/kto/kto_trainer.py [hunk 2]
```diff
@@ -370,15 +593,17 @@ def __init__(
...
+        if not self._is_vision_dataset:
+            train_dataset = self._prepare_dataset(train_dataset, processing_class, args, "train")
+            if eval_dataset is not None:
+                if isinstance(eval_dataset, dict):
+                    eval_dataset = {
+                        key: self._prepare_dataset(dataset, processing_class, args, key)
+                        for key, dataset in eval_dataset.items()
+                    }
+                else:
+                    eval_dataset = self._prepare_dataset(eval_dataset, processing_class, args, "eval")
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py:616]: Vision datasets skip _prepare_dataset entirely for performance reasons: iterating over a large image collection to unpair it would be too expensive. 

Added a cheap first-example check that raises a clear ValueError pointing users to unpair_preference_dataset instead.

**Comment by qgallouedec** [comment]
good call choosing the on-the-fly tokenization. I'm just realizing that we don't have it in DPO (!!!).

File: trl/experimental/kto/kto_trainer.py [hunk 3]
```diff
@@ -158,6 +161,202 @@ def torch_call(self, examples: list[dict[str, Any]]) -> dict[str, Any]:
...
+        return_tensors (`str`, *optional*, defaults to `"pt"`):
+            The tensor type to return. Only `"pt"` is supported.
+    """
+
+    processor: ProcessorMixin
+    max_length: int | None = None
+    calculate_kl: bool = True
+    return_tensors: str = "pt"
+
+    def torch_call(self, examples: list[dict[str, Any]]) -> dict[str, Any]:
```
Comment (qgallouedec) [inline trl/experimental/kto/kto_trainer.py:204]: I would be nice if we could further align the `trl.trainer.sft_trainer.DataCollatorForVisionLanguageModeling._collate_prompt_completion` with this method. Eg, you use `seq_mask` here, but the former uses `completion_mask` (which is imo more accurate/standard); some comments diverge; etc 

It's acceptable to modify `trl.trainer.sft_trainer.DataCollatorForVisionLanguageModeling._collate_prompt_completion`, the only thing that matters here is consistency

**Comment by albertvillanova** [comment]
> good call choosing the on-the-fly tokenization. I'm just realizing that we don't have it in DPO (!!!).

But we already have  on-the-fly tokenization for VLMs in DPO. Or am I missing something?

**Comment by qgallouedec** [comment]
oh wow, I was on a super old branch 🤦‍♂️

File: trl/experimental/kto/kto_trainer.py [hunk 4]
```diff
@@ -158,6 +161,208 @@ def torch_call(self, examples: list[dict[str, Any]]) -> dict[str, Any]:
...
+            for example in examples:
+                example["images"] = [example.pop("image")]
+
+        images = [example["images"] for example in examples]
+        if all(img_list == [] for img_list in images):
+            images = None
+
+        if is_conversational(examples[0]):
+            for example in examples:
+                img_list = example["images"] if images is not None else []
```
Comment (qgallouedec) [inline trl/experimental/kto/kto_trainer.py]: I don't think this is useful:

https://github.com/huggingface/trl/blob/62b234da8e4b37b77d9184de5965b4ca6d6efa61/trl/data_utils.py#L75-L76

File: trl/experimental/kto/kto_trainer.py (hunk 3 above)
Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py:204]: Done!

**Review by qgallouedec — APPROVED** [review]
Ok I think it looks good!
